### ISSUE TITLE
How do I get the available IP ??

### ISSUE THREAD

**Issue by Durgaprasadala** [infobloxopen/infoblox-client#157]
I have this requirement to check if the last octet is .1/.2/.3/.4 of the CIDR and objects.IPAllocation.next_available_ip_from_cidr is not returning the IP.

How do I get the next available IP without allocating from API call like below
self.baseURL + 'network/%s?_function=next_available_ip' % network_ref_id

**Comment by yuewko** [comment]
> I have this requirement to check if the last octet is .1/.2/.3/.4 of the CIDR and objects.IPAllocation.next_available_ip_from_cidr is not returning the IP.

So, the above doesn't work? Could you please attach the code and indicate what result you are getting?

Thanks!

**Comment by Durgaprasadala** [comment]
DEBUG:requests.packages.urllib3.util.retry:Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None)
IPv4: configure_for_dhcp="False", ip="func:nextavailableip:10.65.11.0/26,default", mac="28:92:4a:34:70:68", ipv4addr="func:nextavailableip:10.65.11.0/26,default"
func:nextavailableip:10.65.11.0/26,default

**Comment by Durgaprasadala** [comment]
ip = objects.IPAllocation.next_available_ip_from_cidr('default', cidr)

I want to get the available IP and getting the string "func:nextavailableip:10.65.11.0/26,default" is not helping

**Comment by yuewko** [comment]
> ip = objects.IPAllocation.next_available_ip_from_cidr('default', cidr)

Oh, I see....
You should be using the methods in object_manager.py, e.g. create_fixed_address_from_range().

(The IPAllocation method you're using are lower level methods used by object_manager.py)

**Comment by Durgaprasadala** [comment]
That doesn't work :(
Let me rephrase.

Can I get the next available IP without creating the record(host or A) from the given CIDR.
I would like to take action as per the available IP address( by checking the last octet and decide whether to make it fixed adress or create host record.

Currently, I have to create the host record which will provide the IP address after creating it, then take action as per the last octet.

I just want the next available IP..

**Comment by yuewko** [comment]
If you use create_**fixed_address**_from_cidr(), then only an IP will be allocated without creating any Host or A record.

Unless I'm still missing something...

**Comment by Durgaprasadala** [comment]
Thanks a lot for helping me.

I am using that logic but, it takes a lot of code to achieve that.

1> Create the fixed address with mac=X
2>if the output IP address last octet  is .1/.2/.3/.4 , I cannot leave the fixed address and move to the next available IP as the MAC conflicts.
3>need to write a lot of code(4 times more) to achieve this as mac address conflict issue is there.

Is there any way I can create the fixed address without providing "mac" ???

**Comment by yuewko** [comment]
You can provide `"00:00:00:00:00:00"` as the MAC address. Please give that a try.

**Comment by yuewko** [comment]
BTW, is it your goal to skip .1 to .4 when allocating IP? If that's the case, you can use create_fixed_address_from_range(), which lets you to specify the range.

**Comment by Durgaprasadala** [comment]
For some reason create_fixed_address_from_range() is not going to work properly for me as the records are scattered.

I am almost there with the solution. need a little bit help.

1>Create the fixed address with MAC='00:00..' , that reserved the IP.
2>Now I am trying to create the host-record on top of reserved IP I just got.
infoblox_client.exceptions.InfobloxCannotCreateObject: Cannot create 'record:host' object(s): { "Error": "AdmConDataError: None (IBDataConflictError: IB.Data.Conflict:IP address 10.65.11.4 is already used by a reservation.)",

How can I convert the reservation into host record without deleting( another step).

**Comment by Durgaprasadala** [comment]
I tried create_host_record_from_range , I can provide the first_ip starting from .5 and can I provide the wild card for last_ip ???

It is not working for me .

**Comment by Durgaprasadala** [comment]
This is what I am doing it right now, How to optimize it ?

0>Create the fixed IP with MAC="00:00.."  ,reserved the IP
Check if IP matches .1 or .2 or .3 or .4

1> If it matches : --->   do step 0>

2>If it doesn't match: ---> delete the fixed and create the host record.

Not happy with the solution, but live with it.

the last resort is :::: I will make a direct api call to get the next available IP.
network/%s?_function=next_available_ip' % network_ref_id

A feature to get the next available IP without creating any object will awesome.

**Comment by alekgrom** [comment]
How is it possible using this library to get next available IP excluding a certain list of IP addresses? 
According to documentation to POST body should be added **exclude** variable.
```
URL:
https://1.2.3.4/wapi/v2.6/network/
ZG5zLm5ldHdvcmskMTAuMC4wLjAvMjQvMA:10.0.0.0/24/default?
_function=next_available_ip
DATA:
’exclude’: [’9.0.0.1’, ’9.0.0.8’]
```
